### ISSUE TITLE
TWCS: always keeps expired sstables for additional gc_grace_seconds after they expire

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#9833]
*Installation details*
HEAD: b2a62d2b59a97bbf2e3254a5b50bb935d35b3851

*Description*
TWCS doesn't consider sstables for eviction till

```
max_deletion_time < now() - gc_grace_seconds (=gc_before)  (*)
```

However for TTLed data this means that sstables won't be considered for eviction till:
```
max_write_time + TTL < now() - gc_grace_second  (**)
```

However for TTLed data gc_grace is not counted since the "deletion time" but rather from the write time (read https://docs.scylladb.com/kb/ttl-facts/ for more details).
And therefore the generic compaction logic (*) above is going to keep completely expired sstables for gc_grace_second for no reason (when TTL is greater than gc_grace_seconds - which is the case here) for a data that is using TTL.

*How to reproduce*
1) Create a single node Scylla cluster (e.g. using `ccm`).
2) Make cassandra-stress populate some data (you only need it to create its schema).
3) Using cqlsh:
  a) Truncate the `keyspace1.standard1`
  b) Set it's compaction strategy and default TTL and gc_grace_seconds as follows:
```
cqlsh> alter table keyspace1.standard1 with compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '1', 'compaction_window_unit': 'MINUTES'};
cqlsh> alter table keyspace1.standard1 with default_time_to_live = 180;
cqlsh> alter table keyspace1.standard1 with gc_grace_seconds = 120;
```
4) Run cassandra-stress again:
```
cassandra-stress write n=900000000000 no-warmup -rate threads=300 throttle=10000/s -mode native cql3 user=cassandra password=cassandra -pop seq=1..900000000000
```
5) Let it run for 6 or more minutes and see that there many more time window buckets than expected (5).

**Comment by vladzcloudius** [comment]
We should probably consider extend a generic logic to consider TTL info when deciding to consider an sstable to be a "expired" candidate.

For instance if we see that minTTL > 0 (meaning that all data in this sstable is TTLed) we should consider the sstable for expiration candidate using the following logic:

```
if gc_grace_seconds > minTTL:
   <use the current logic>
elif maxWriteTime + minTTL < now() and !has_undeleted_ancestor(candidate):
   <consider this sstable for expired candidates>
```

**Comment by vladzcloudius** [comment]
@eliransin @raphaelsc FYI

**Comment by bhalevy** [comment]
@raphaelsc let's reproduce that with a dtest that will verify the fix

**Comment by slivne** [comment]
I fail to understand - data needs to be kept for gc_grace_seconds after TTL - we can't "cut" this down. If data is retained for longer then we can check it out as to why (if compaction is not triggered - sstable may stay). 

ssatble have to stay for gc_grace_seconds after everthing has become a tombstone we should not trim that down.

If anyone want to trim this down he can set a different gc_grace_seconds.

**Comment by bhalevy** [comment]
The reason gc_grace_seconds exists is to let compaction and repair time to spread tombstones in the clusters and make they are compacted with and data that shadow, then they can be purged when we make sure they do not possibly shadow and data in local sstables.

With TTLed data, the "gc_grace" period starts ticking when the data is written, not when it becomes a tombstone, since the same logic applies to the TTLed cells, where repair and compaction need to make sure they are spread around in the cluster and be compacted with any data they shadow, even when alive and they have gc_grace_seconds to do so from when the TTLed data was *written*, not from when it became a tombstone.

When the TTL expires, if gc_grace_seconds had already passed since the time it was written, it can be purged immediately, modulo checking it doesn't possibly shadow data in other sstables, exactly like we do for gc-expired tombstones.

**Comment by vladzcloudius** [comment]
> I fail to understand - data needs to be kept for gc_grace_seconds after TTL - we can't "cut" this down.

@slivne you are wrong. Data doesn't need to be retained for gc_grace after TTL but rather after a write. See the Document I referenced in the opening message - it's called "Not so obvious facts about TTL" for a reason ;).

**Comment by slivne** [comment]
The assumption in the above is that the user DOES NOT use cql writetime to modify the writetime of an element - we have no such grantee. 

So a user can write an element with CQL writetime and we will not have enough time to replicate the data (and its TTL) in the cluster and need to rely on the gc_grace_seconds for the tombstone.

Till we have something blocking the use of CQL WRITETIME you can't base any logic on it not being used.

**Comment by slivne** [comment]
@bhalevy / @vladzcloudius ^^

**Comment by raphaelsc** [comment]
> @bhalevy / @vladzcloudius ^^

in the low-level mutation compaction code, we already use write time when determining if a TTLd cell can be removed
- https://github.com/scylladb/scylla/blob/542394c82f01ff522e4128b0a115b851121a4f95/mutation_partition.cc#L1578-L1580
- https://github.com/scylladb/scylla/blob/542394c82f01ff522e4128b0a115b851121a4f95/atomic_cell.hh#L264

Note how deletion_time definition for TTLd cell is: expiry() - ttl() (expiry = expiration time, so expiry() - ttl = cell's write time)

with table that has default TTL = gc period = 60, major compaction can purge away all expired data after 60 seconds, not after 120. I have confirmed that by running a manual test under the described scenario.

assuming current behavior is correct, Vlad proposal makes sense as it's just adjusting high level compaction code to work with the same definition of deletion_time for TTLd data as low-level mutation compaction mechanism.

**Comment by vladzcloudius** [comment]
> The assumption in the above is that the user DOES NOT use cql writetime to modify the writetime of an element - we have no such grantee.

@slivne You are confused: there are no "assumptions" in what I described - these are facts and they are as they are whether you like them or not.

The problem is in the generic compaction fan-out logic that doesn't take the TTL expiration calculation and its relation with gc_grace into an account.

As to your comment about 'USING TIMESTAMP' its relation to TTL is also not trivial and is described in a "Note" section here: https://docs.scylladb.com/getting-started/time-to-live/

Shlomi, please, read the documentation pages I'm referencing before posting the next comment.

**Comment by slivne** [comment]
Ok - found my mistake - the expiry is computed since now() thanks not the writetime. 

Having said that, I fail to understand how this is related specifically to TWCS

Its a general question should TTL data after its converted into a tombstone be retained for an additional gc_grace_seconds or not - at least this is what cassandra does - tested on cassandra 3.11.3

I think this is the original issue coming from cassandra https://issues.apache.org/jira/browse/CASSANDRA-5546 - there is a backward compatibility issue here - and that means we suffer from it as well - e.g. users that have used tables with default_time_to_live + low gc_grace - assume that repair will run in a window of default_time_to_live + gc_grace not in gc_grace_seconds (or so it seems based on the cassandra issue dated from 2016).

**Comment by raphaelsc** [comment]
> Ok - found my mistake - the expiry is computed since now() not the writetime thanks.
> 
> Having said that, I fail to understand how this is related specifically to TWCS

@slivne, [as I explained above](https://github.com/scylladb/scylla/issues/9833#issuecomment-1000445525), compaction is already using now() concept (expiration time - TTL) when purging expired data. If you force major on expired SSTable, you will see compaction is able to purge all expired data, provided its TTL > gc period. TWCS, however, has an optimization that bypasses compaction for fully expired sstables, but this optimization is based on expiration (deletion) time. 

So given the fact that we already use TTLd data's now() in compaction, it makes sense to adjust TWCS too, so expired sstables containing only TTLd data don't have to wait for additional gc grace seconds.

Many (most?) TWCS users set gc period to 0, so this optimization makes sense for users which perform (bulk) deletions as they have to set gc grace seconds to some significant value.

**Comment by vladzcloudius** [comment]
> Many (most?) TWCS users set gc period to 0, so this optimization makes sense for users which perform (bulk) deletions as they have to set gc grace seconds to some significant value.

@raphaelsc While I agree with all your points from the previous post - this one requires a correction: setting gc_grace_seconds to 0 is dangerous is you use HH because it would effectively make HH drop every hint and not send it out virtually disabling HH.

So, the correct statement would be: "All TWCS uses should set gc_grace_seconds to reasonably low value if they use HH."

TWCS is *meant* to be used with data that is neither explicitly deleted nor updated. In such a case data resurrection is impossible hence there is no gc_grace-related repair restriction doesn't apply.

For those that DO update or explicitly delete their data and use TWCS the regular restrictions apply: they should run a repair at least once every gc_grace_seconds (yes, the is a "twist" with TTL but it's simpler not to overcomplicate the condition).

> Its a general question should TTL data after its converted into a tombstone be retained for an additional gc_grace_seconds or not - at least this is what cassandra does - tested on cassandra 3.11.3

@slivne You are right and this relates to any compaction that needs to deal with TTLed data.
My understanding was that what's described in  https://docs.scylladb.com/kb/ttl-facts/  IS a correct behavior and the JIRA you referenced above confirms that. 
I don't think there is any backward compatibility issue since we have documented this behavior and it's in our implementation for quite some time (from day 0 AFAIU). So, let's move on and make our compactions behavior consistent like I've described in the opening message.
